[10063] Update versions of mypy and co to work with pip.

adiroiban opened the pull request:
This is an updated of mypy so that there are no conflicts between pinned version.

It updates the mypy version as I guess we will want to upgrade at some point.

It updates the twisted own code to not trigger any mypy errors.


## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10063
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I ran additional checks listed at [Getting Your Patch Accepted](https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] I have updated the automated tests.
* [ ] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

adiroiban committed:
Initial mypy update.

graingert commented on src/twisted/application/twist/test/test_options.py:
```diff
@@ -47,7 +47,7 @@ def patchOpen(self) -> None:
         """
         self.opened = []  # type: List[Tuple[str, Optional[str]]]
 
-        def fakeOpen(name: str, mode: Optional[str] = None) -> NotImplemented:
```
this doesn't return Exception

adiroiban replied:
What is your suggestion for fixing this? :) thanks

mthuurne replied:
I wanted to suggest `Literal[NotImplemented]`, but mypy doesn't like that either:
```
error: Parameter 1 of Literal[...] is invalid  [misc]
```

mthuurne replied:
You could use `object` as the return type, that is both correct and will make mypy complain if the returned value is used as another type.

graingert commented on src/twisted/python/runtime.py:
```diff
@@ -108,7 +108,7 @@ def isVista(self) -> bool:
         @return: C{True} if the current platform has been detected as Vista
         """
         if getattr(sys, "getwindowsversion", None) is not None:
```
```
windowsversion = getattr(sys, "getwindowsversion", None)
return isinstance(windowsversion, tuple) and windowsversion[0] == 6
```

adiroiban replied:
Is it worth doing this `isinstance` check instead of just ignoring it ?

mthuurne replied:
That suggestion would work around the `attr-defined` warning but not the `no-any-return` warning, right?

graingert replied:
It's certainly odd to lookup the attribute twice

graingert replied:
```
windowsversion = getattr(sys, "getwindowsversion", None)
return bool(isinstance(windowsversion, tuple) and windowsversion[0] == 6)
```

mthuurne replied:
I'd say inefficient, but not odd, since it happens a lot in code.

Actually, that suggestion is incorrect, since `getwindowsversion` is a function, not a tuple.

graingert replied:
@mthuurne ah of course, I think this should typecheck:

`return sys.platform == "win32" and sys.getwindowsversion().major == 6`

graingert replied:
yep: https://mypy-play.net/?mypy=latest&python=3.6&flags=strict&gist=bca7bed5e8445b6e8e9e10a1f2aec229

mthuurne commented on tox.ini:
```diff
@@ -139,9 +139,9 @@ deps=https://github.com/twisted/pydoctor/archive/3f9c64829dfa040b334c9ae27c332c7
 deps =
-    mypy==0.782
+    mypy==0.790
     mypy_extensions==0.4.3
```
```suggestion
```

mthuurne commented on src/twisted/python/runtime.py:
```diff
@@ -108,7 +108,7 @@ def isVista(self) -> bool:
         @return: C{True} if the current platform has been detected as Vista
         """
         if getattr(sys, "getwindowsversion", None) is not None:
```
```suggestion
        if hasattr(sys, "getwindowsversion"):
```

graingert replied:
```suggestion
        return sys.platform == "win32" and sys.getwindowsversion().major == 6
```

graingert commented on src/twisted/application/twist/test/test_options.py:
```diff
@@ -47,7 +47,7 @@ def patchOpen(self) -> None:
         self.opened = []  # type: List[Tuple[str, Optional[str]]]
 
-        def fakeOpen(name: str, mode: Optional[str] = None) -> NotImplemented:
+        def fakeOpen(name: str, mode: Optional[str] = None) -> Exception:
```
```suggestion
        def fakeOpen(name: str, mode: Optional[str] = None) -> object:
```

https://github.com/python/mypy/issues/363

adiroiban committed:
Remove the exensions.

adiroiban commented:
Thanks for the feedback.

I don't plan to continue working on this.
Feel free to commit your suggestions. I will be happy to review them.

Thanks!

graingert commented on src/twisted/python/runtime.py:
```diff
@@ -108,7 +108,7 @@ def isVista(self) -> bool:
         """
         if getattr(sys, "getwindowsversion", None) is not None:
-            return sys.getwindowsversion()[0] == 6
+            return sys.getwindowsversion()[0] == 6  # type: ignore[no-any-return, attr-defined]
         else:
             return False
```
```suggestion
```

graingert committed:
check sys.platform before calling sys.getwindowsversion()

to appease mypy

graingert committed:
type annotate fakeOpen to return object

because NoReturn is an alias for Any https://github.com/python/mypy/issues/363

graingert committed:
handle sorting None and str subCommands

src/twisted/application/twist/_options.py:187:21: error: Value of type variable "_LT" of "sorted" cannot be "Optional[str]"  [type-var]
            for name in sorted(plugins):

mthuurne committed:
Remove Optional[] from plugin name annotation

In practice the name is never None, but we need to express that in
a way that mypy understands.

mthuurne reviewed (approved):
Looks good, but now we have to figure out what makes `pip` crash so we can get CI to pass.

adiroiban commented on src/twisted/application/twist/_twist.py:
```diff
@@ -102,8 +102,12 @@ def main(cls, argv: Sequence[str] = sys.argv) -> None:
+        # If subCommand is None, TwistOptions.parseOptions() raises UsageError
+        # and Twist.options() will exit the runner, so we'll never get here.
+        subCommand = options.subCommand
+        assert subCommand is not None
```
I think that this needs an inline comment to note that this line was added to make mypy happy... as if you remove it, mypy wil fail.

```suggestion
        assert subCommand is not None  # Here to help mypy.
```

graingert replied:
I think the comment is covered in 105-106, also it's very idiomatic to use an `assert x is not None` to unpack an Optional

adiroiban commented on src/twisted/application/twist/test/test_twist.py:
```diff
@@ -108,8 +108,10 @@ def test_startService(self) -> None:
 
         reactor = options["reactor"]
+        subCommand = options.subCommand
+        assert subCommand is not None
```
Maybe inline commnent about mypy ?

adiroiban reviewed (commented):
All good. Approved. Fenomenal work :)

I think that is best to merge this as it is and fix the docs in https://github.com/twisted/twisted/pull/1492

adiroiban commented:
Thanks. I will merge it as it is.
